tests/pipeline/test_slice.py

```diff
@@ -0,0 +1,304 @@
+        with self.assertRaises(UnsupportedPipelineColumn):
+            pipe.add(open_slice, 'open_slice')
+
+    def test_loadable_term_slices(self):
+        """
+        Test that slicing loadable terms raises the proper error.
+        """
+        my_asset = self.asset_finder.retrieve_asset(self.sids[0])
+
+        with self.assertRaises(NonSliceableTerm):
+            USEquityPricing.close[my_asset]
+
+    def test_non_existent_asset(self):
+        """
+        Test that indexing into a term with a non-existent asset raises the
+        proper exception.
+        """
+        my_asset = Asset(0)
+        returns = Returns(window_length=2)
+        returns_slice = returns[my_asset]
+
+        class UsesSlicedInput(CustomFactor):
+            window_length = 1
+            inputs = [returns_slice]
+
+            def compute(self, today, assets, out, returns_slice):
+                pass
+
+        with self.assertRaises(NonExistentAssetInTimeFrame):
+            self.engine.run_pipeline(
+                Pipeline(columns={'uses_sliced_input': UsesSlicedInput()}),
+                self.start_date,
+                self.end_date,
+            )
+
+    def test_window_safety_of_slices(self):
+        """
+        Test that slices correctly inherit the `window_safe` property of the
+        term from which they are derived.
+        """
+        my_asset = self.asset_finder.retrieve_asset(self.sids[0])
+
+        # SimpleMovingAverage is not window safe.
+        sma = SimpleMovingAverage(
+            inputs=[USEquityPricing.close], window_length=10,
+        )
+        sma_slice = sma[my_asset]
+
+        class UsesSlicedInput(CustomFactor):
+            window_length = 1
+            inputs = [sma_slice]
+
+            def compute(self, today, assets, out, sma_slice):
+                pass
+
+        with self.assertRaises(NonWindowSafeInput):
+            self.engine.run_pipeline(
+                Pipeline(columns={'uses_sliced_input': UsesSlicedInput()}),
+                self.start_date,
+                self.end_date,
+            )
+
+        # Make sure that slices of custom factors are not window safe.
+        class MyFactor(CustomFactor):
+            window_length = 1
+            inputs = [USEquityPricing.close]
+
+            def compute(self, today, assets, out, close):
+                out[:] = close
+
+        my_factor = MyFactor()
+        my_factor_slice = my_factor[my_asset]
+
+        class UsesSlicedInput(CustomFactor):
+            window_length = 1
+            inputs = [my_factor_slice]
+
+            def compute(self, today, assets, out, my_factor_slice):
+                pass
+
+        with self.assertRaises(NonWindowSafeInput):
+            self.engine.run_pipeline(
+                Pipeline(columns={'uses_sliced_input': UsesSlicedInput()}),
+                self.start_date,
+                self.end_date,
+            )
+
+        # Slices of Returns *are* window safe.
+        returns = Returns(window_length=2)
```

Comment:
For the purpose of this test, I think we care more about the fact that the term here is window_safe than that it's actually `Returns`.  I would probably make a new `MySafeFactor` here with window_safe=True and test that.